courses/fundamentals_of_ada/adv_090_overloading_equality.rst

```diff
@@ -0,0 +1,269 @@
+
+------------------------------------
+User-Defined `=` Returning Boolean
+------------------------------------
+
+* Implicitly declares ``/=``
+* Thus negation has consistent meaning
+
+   .. code:: Ada
+
+      if X /= Y then
+      if not ( X = Y ) then
+      if X not = Y then
+
+* No explicit declaration of ``/=`` returning Boolean
+
+   - Returning values of other types is allowed
+
+      .. code:: Ada
+
+         function "/=" (Left : Foo;  Right : Bar)
+             return Fuzzy_Result;
+
+-------------------------------
+User-Defined Equality Example
+-------------------------------
+
+* Especially useful for composite types
+* Predefined ``=`` is bit-wise comparison over entire structure so may be inappropriate semantics
+* Given the following types:
+
+   .. code:: Ada
+
+      Max : constant := 100;
+      type Index is range 0 .. Max;
+      type List is array (Index range 1 .. Max) of Integer;
+      type Stack is record
+        Values : List;
+        Top : Index := 0;
+      end record;
+
+* Equality function might look like:
+
+   .. code:: Ada
+
+      function "=" (Left, Right : Stack) return Boolean is
+      begin
+        if Left.Top /= Right.Top then -- not same size
+          return False;
+        else -- compare values
+          for K in 1 .. Left.Top loop
+            if Left.Values(K) /= Right.Values(K) then
+              return False;
+            end if;
+          end loop;
+        end if;
+        return True;
+      end "=";
+
+=========================
+Composition of Equality
+=========================
+
+----------
+Examples
+----------
+
+.. include:: examples/090_overloading/composition_of_equality.rst
+
+----------------------------
+ "Composition of Equality"
+----------------------------
+
+* Whether user-defined equality functions are called automatically as part of equality for composite types containing types having such functions
+* Only composes when user-defined equality is defined
+
+   * Assume you defined "=" for a scalar type
+   * If you define "=" for a composite containing the scalar type, your scalar "=" will be used
+   * If you rely on the implicit "=" for the composite, then the scalar's implicit "=" will also be used
+
+      * Not the one you just defined
+
+--------------------------------
+Composition vs Non-Composition
+--------------------------------
+
+.. code:: Ada
+
+   with Ada.Text_IO; use Ada.Text_IO;
+   procedure Main is
+
+      type Array1_T is array (1 .. 3) of Integer;
+      type Array2_T is array (1 .. 3) of Integer;
+
+      X, Y     : Integer  := 123;
+      X_A, Y_A : Array1_T := (others => 123);
+      X_B, Y_B : Array2_T := (others => 123);
+
+      -- When comparing integers directly, this function forces those comparisons
+      -- to be false
+      function "=" (L, R : Integer) return Boolean is
+      begin
+         return False;
+      end "=";
+      -- We define our own array equality operator so it will use our integer operator
```

Comment:
That was ported as it was written, the error was part of the original material.

I believe the idea is that we defined our own composite type (array) equality, so in its body it will use the user-defined integer equality instead of the default one.

Reply:
my comment wasn't clear - "integer operator" should be "integer equality operator" (the second part of the sentence, not the first part)

Reply:
Done, this section has been moved to advanced on master, I'll push the PR for the fix soon

Reply:
Done :)